SSA/Core/EDSLNested.lean

```diff
@@ -0,0 +1,257 @@
+import SSA.Core.WellTypedFramework
+/-
+an EDSL that elaborates expressions with nesting
+allowed into SSA.
+
+Authors: Siddharth Bhat
+-/
+namespace EDSL2
+open Lean Elab Macro
+
+declare_syntax_cat dsl_op2
+scoped syntax "[dsl_op2|" dsl_op2 "]" : term
+syntax dsl_var2 := "%v" num
+declare_syntax_cat dsl_expr2
+declare_syntax_cat dsl_region2
+declare_syntax_cat dsl_stmt2
+
+/-
+This is a fancier syntax, which allows expressions in the
+intermediate nodes. This is elaborated into SSA, with temporary
+variable names.
+-/
+syntax dsl_var2 : dsl_expr2
+scoped syntax "()" : dsl_expr2
+scoped syntax "(" dsl_expr2 ")" : dsl_expr2
+scoped syntax "(" dsl_expr2 "," dsl_expr2 ")" : dsl_expr2
+scoped syntax "(" dsl_expr2 "," dsl_expr2 "," dsl_expr2 ")" : dsl_expr2
+scoped syntax "op:" dsl_op2 ws dsl_expr2 (ws dsl_region2)? : dsl_expr2
+syntax dsl_assign2 := dsl_var2 ":= " dsl_expr2
+scoped syntax sepBy(dsl_assign2, ";") : dsl_stmt2
+
+syntax dsl_bb2 := (dsl_stmt2)? "return " dsl_expr2
+
+scoped syntax "{" dsl_var2 "=>" dsl_bb2 "}" : dsl_region2
+-- scoped syntax "$(" term ")" : dsl_region2
+-- scoped syntax "$(" term ")" : dsl_expr2
+
+inductive ElabVar
+| User (n : ℕ)
+| Synthetic (n : ℕ)
+deriving DecidableEq, Inhabited
+
+def ElabVar.toString : ElabVar → String
+| .User n => s!"%v{n}"
+| .Synthetic n => s!"%vsynth{n}"
+
+instance : ToString ElabVar where toString := ElabVar.toString
+
+def ElabVar.toNat : ElabVar → Nat
+| .User n => n
+| .Synthetic n => n
+
+def ElabVar.quoteAsNat (var : ElabVar) : TSyntax `term := 
+  Lean.quote var.toNat
+
+
+
+structure SSAElabContext where
+  vars : Array ElabVar -- list of variables, in order of occurrence
+
+abbrev SSAElabM (α : Type) := StateT SSAElabContext MacroM α
+
+def SSAElabContext.addVar (var : ElabVar) : SSAElabM Unit := do
+  modify fun ctx => { ctx with vars := ctx.vars.push var }
+
+/- generate a fresh variable for temporary use. -/
+def SSAElabContext.addFreshSyntheticVar : SSAElabM ElabVar := do
+  -- This is a hack. We assign fresh variables an index > 99999.
```

Comment:
Why not make a `def SynthVarIndex : Nat -> Nat -> Nat := fun prefix var => prefix + var` and at the end of the elaboration you just instantiate them all with the number of `User` vars? (Probably you'd want to make it a bit nicer than this, e.g. my making your `SSAElabContext` carry the two scopes and then the synthetic variables are parametrized by a single prefix there or so, such that your partially generated syntax is all parametrized by the same prefix)`

Reply:
Unfortunately, it does not suffice to have `prefix + var`, as the synthetic variables must not clash with _any_ variable (even those that occur later!). Thus, the macro expansion would need to be two pass (one pass to collected the largest index that a user uses, and a second pass to fixup the index for all synthetic variables).

Reply:
@bollu yeah I know, that's exactly why I described it like this, as a *parameter* rather than just adding the (constant) prefix up to that point. If it's a parameter, there's no need to do two passes, you can just instantiate the parameter in the end after the first pass with 1 + the greatest user variable index.

Reply:
@goens to be frank, I still I don't see how to do this in a single pass. I'm unsure how the "fixup" is supposed to be done. 

If I were in Haskell, I could express it in a single pass via [Tardis](https://hackage.haskell.org/package/tardis-0.4.4.0/docs/Control-Monad-Tardis.html) and laziness to pass the final value "back up". Here, I'm not too sure 😄 

I propose that you take a stab at it once this is merged. Does that sound like a fair compromise?